### ISSUE TITLE
Provide a way to set the config file include prefix to OPENSSLDIR

### ISSUE THREAD

**Issue by qunying** [openssl/openssl#17704]
To test the fips module, I follow the guide line to generate the fipsmodule.cnf and try to use it in openssl.cnf. But it fails with error:
    ```
   missing config data
    ```
If I change the include line to
    ```
   .include /etc/ssl/fipsmodule.cnf
    ```
Then it will work.
 ```
openssl version -a
OpenSSL 3.0.1 14 Dec 2021 (Library: OpenSSL 3.0.1 14 Dec 2021)
built on: Sat Feb 12 02:33:00 2022 UTC
platform: linux-x86_64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
OPENSSLDIR: "/etc/ssl"
ENGINESDIR: "/usr/lib64/engines-3"
MODULESDIR: "/usr/lib64/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_ia32cap=0xe3bdefebffff:0x0
```
I tried to set the `HOME = /etc/ssl` without success.
A simple test program I used to just load the fips providers to test:
```
#include <stdio.h>
#include <openssl/provider.h>
#include <openssl/err.h>

int
main(int argc, char *argv[])
{
    OSSL_PROVIDER *fips_provider;

    fips_provider = OSSL_PROVIDER_load(NULL, "fips");
    if (fips_provider == NULL) {
        unsigned long err = ERR_get_error();
        printf("fips provider load failed: %s\n", ERR_reason_error_string(err));
        return 1;
    }

    return 0;
}
```
Working configuration is:
[openssl.cnf.txt](https://github.com/openssl/openssl/files/8063362/openssl.cnf.txt)

**Comment by mattcaswell** [comment]
You show us what your working configuration looks like, but not what you tried that was not working. I am guessing you were using a relative pathname, e.g. something like this:

````
.include fipsmodule.cnf
````

Is that correct?

The config man page describes how such relative filenames are processed:

````
As a general rule, the pathname should be an absolute path; this can be enforced with the abspath and includedir pragmas, described below. The environment variable OPENSSL_CONF_INCLUDE, if it exists, is prepended to all relative pathnames. If the pathname is still relative, it is interpreted based on the current working directory.
````
https://www.openssl.org/docs/man3.0/man5/config.html


> I follow the guide line

Were you following a specific man page or guide that led you to do this? If so which one. The fips module man page seems use absolute paths:

https://www.openssl.org/docs/man3.0/man7/fips_module.html

**Comment by qunying** [comment]
It would be better if the default sample` openssl.cnf` include the full path for the `.include fipsmodule.cnf` line. As I only read through the sample configuration file and un-comment the line that include the fipsmodule line and expecting it to work.

**Comment by t8m** [comment]
The sample file is used for tests and that means it cannot include full path otherwise it would have to be mangled. However we might add a comment.

**Comment by mattcaswell** [comment]
Adding a comment would be a good idea.

**Comment by CristiFati** [comment]
On *v3.0.8* (didn't yet build *v.3.1.0*, but I assume it's the same thing) I found 2 workarounds (out of which (at least) one is lame (*gainarie*)):

- Change current directory to *OPENSSLDIR*

- Set *OPENSSL\_CONF\_INCLUDE* *env* *var* to *OPENSSLDIR*

```
(py_pc064_03.10_test1_pycoral) [cfati@cfati-5510-0:/usr/local/pc064/openssl/openssl/3.0.8]> ~/sopr.sh
### Set shorter prompt to better fit when pasted in StackOverflow (or other) pages ###

[064bit prompt]> ls
bin  include  lib  share  ssl
[064bit prompt]>
[064bit prompt]> ./bin/openssl md5 ./bin/c_rehash
FATAL: Startup failure (dev note: apps_startup()) for ./bin/openssl
8012C0065D7F0000:error:80000002:system library:process_include:No such file or directory:crypto/conf/conf_def.c:805:calling stat(fipsmodule.cnf)
8012C0065D7F0000:error:07800069:common libcrypto routines:provider_conf_load:provider section error:crypto/provider_conf.c:156:section=fips_sect not found
8012C0065D7F0000:error:0700006D:configuration file routines:module_run:module initialization error:crypto/conf/conf_mod.c:270:module=providers, value=provider_sect retcode=-1
[064bit prompt]>
[064bit prompt]> pushd ./ssl; ../bin/openssl md5 ../bin/c_rehash; popd
/usr/local/pc064/openssl/openssl/3.0.8/ssl /usr/local/pc064/openssl/openssl/3.0.8
Error setting digest
8092D4D5987F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (MD5 : 102), Properties ()
8092D4D5987F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:254:
/usr/local/pc064/openssl/openssl/3.0.8
[064bit prompt]>
[064bit prompt]> OPENSSL_CONF_INCLUDE=./ssl ./bin/openssl md5 ./bin/c_rehash
Error setting digest
80427A6D667F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (MD5 : 102), Properties ()
80427A6D667F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:254:
[064bit prompt]>
[064bit prompt]> OPENSSL_CONF_INCLUDE=./ssl ./bin/openssl version -a
OpenSSL 3.0.8 7 Feb 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)
built on: Sun Apr  2 09:04:43 2023 UTC
platform: linux-x86_64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
OPENSSLDIR: "/usr/local/pc064/openssl/openssl/3.0.8/ssl"
ENGINESDIR: "/usr/local/pc064/openssl/openssl/3.0.8/lib/engines-3"
MODULESDIR: "/usr/local/pc064/openssl/openssl/3.0.8/lib/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_ia32cap=0xfedab2035f8bffff:0x9c2fb9
```

**Comment by junaruga** [comment]
Maybe I see this issue in my case with a testing program (fips_mode.c) in the <https://github.com/junaruga/openssl-test> in OpenSSL 3.0.8.

```
$ gcc \
  -I /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/include \
  -L /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/lib64 \
  -lcrypto \
  -o fips_mode \
  fips_mode.c
```

The following OpenSSL config file with the full path `.include /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/fipsmodule.cnf` can load the `fipsmodule.cnf`.

```
$ cat /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf
config_diagnostics = 1
openssl_conf = openssl_init

.include /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/fipsmodule.cnf
#.include ./fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
tfips = fips_sect
base = base_sect

[base_sect]
activate = 1

[algorithm_sect]
default_properties = fips=yes
```

```
$ LD_LIBRARY_PATH=$HOME/.local/openssl-3.0.8-fips-debug-trace/lib64/ \
  OPENSSL_CONF=/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf \
  ./fips_mode
Loaded providers:
  fips
  base
FIPS mode enabled: 1
```

However, the OpenSSL config file with the relative patch `.include ./fipsmodule.cnf` can't load the `fipsmodule.cnf`. I want to see that the OpenSSL config file loads the `.include ./fipsmodule.cnf` as a relative patch from the `/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl` directory.

```
$ cat /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf
config_diagnostics = 1
openssl_conf = openssl_init

#.include /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/fipsmodule.cnf
.include ./fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

[algorithm_sect]
default_properties = fips=yes
```

```
$ LD_LIBRARY_PATH=$HOME/.local/openssl-3.0.8-fips-debug-trace/lib64/ \
  OPENSSL_CONF=/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf \
  ./fips_mode
Loaded providers:
  default
FIPS mode enabled: 0
```

**Comment by junaruga** [comment]
> 1. Change current directory to OPENSSLDIR (as displayed by: openssl version -a)

Thank you for checking it. Interesting. I was able to load the `.include fipsmodule.cnf` by changing the current working directory the `/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl`. Shall we change the default directory used to load the `fipsmodule.cnf` from the current working directory to OPENSSLDIR?

```
$ cat /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf
config_diagnostics = 1
openssl_conf = openssl_init

#.include /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/fipsmodule.cnf
.include fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

[algorithm_sect]
default_properties = fips=yes
```

```
$ pwd
/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl

$ LD_LIBRARY_PATH=$HOME/.local/openssl-3.0.8-fips-debug-trace/lib64/ \
  OPENSSL_CONF=/home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl/openssl_fips.cnf \
  ~/git/openssl-test/fips_mode
Loaded providers:
  fips
  base
FIPS mode enabled: 1
```

**Comment by t8m** [comment]
> Shall we change the default directory used to load the `fipsmodule.cnf` from the current working directory to OPENSSLDIR?

We cannot IMO change this in the library. What we could do is to add a pragma to the default config file that sets the prefix to OPENSSLDIR.

**Comment by junaruga** [comment]
> We cannot IMO change this in the library. What we could do is to add a pragma to the default config file that sets the prefix to OPENSSLDIR.

All right. That looks great idea!

**Comment by junaruga** [comment]
> What we could do is to add a pragma to the default config file that sets the prefix to OPENSSLDIR.

@t8m You meant that OpenSSL enables users to set the prefix to OPENSSLDIR like this?

```
$ cat /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/ssl
...
.include @openssldir@/fipsmodule.cnf
...
```

**Comment by t8m** [comment]
Yes, and we would use that in the default config file that is being installed.

**Comment by CristiFati** [comment]
As I explained in my previous comment, there are 2 workarounds which allow using relative paths in *openssl.conf*.

**Comment by junaruga** [comment]
> Yes (or through the includedir pragma) and we would use that in the default config file that is being installed.

@t8m All right. That sounds great. I found the commit of the `includedir` pragma <https://github.com/openssl/openssl/commit/f7050588bc76901e0a147c158e64ac3140dc8bfd>. And I can see it is set in the pkgconfig files.

```
$ pwd
/home/jaruga/.local/openssl-3.0.8-fips-debug-trace

$ grep -r includedir lib64/
lib64/pkgconfig/libssl.pc:includedir=${prefix}/include
lib64/pkgconfig/libssl.pc:Cflags: -I${includedir}
lib64/pkgconfig/libcrypto.pc:includedir=${prefix}/include
lib64/pkgconfig/libcrypto.pc:Cflags: -I${includedir}
lib64/pkgconfig/openssl.pc:includedir=${prefix}/include
grep: lib64/libcrypto.so.3: binary file matches
grep: lib64/libcrypto.a: binary file matches
```

**Comment by junaruga** [comment]
> As I explained in my previous comment, there are 2 workarounds which allow using relative paths in openssl.conf.

@CristiFati Yes, I saw the workaround you explained. However, I wanted to load the `fipsmodule.cnf` without workarounds, as I want to test with a specific OpenSSL by switching it easily. I have 6 testings OpenSSLs on my local.

```
$ ls -dl /home/jaruga/.local/openssl-*
drwxr-xr-x. 7 jaruga jaruga 4096 Mar 17 22:28 /home/jaruga/.local/openssl-3.0.8/
drwxr-xr-x. 7 jaruga jaruga 4096 Apr  1 00:31 /home/jaruga/.local/openssl-3.0.8-debug/
drwxr-xr-x. 7 jaruga jaruga 4096 Apr 28 16:50 /home/jaruga/.local/openssl-3.0.8-debug-trace/
drwxr-xr-x. 7 jaruga jaruga 4096 Mar 13 19:28 /home/jaruga/.local/openssl-3.0.8-fips/
drwxr-xr-x. 7 jaruga jaruga 4096 Mar 28 19:36 /home/jaruga/.local/openssl-3.0.8-fips-debug/
drwxr-xr-x. 7 jaruga jaruga 4096 Apr 25 16:44 /home/jaruga/.local/openssl-3.0.8-fips-debug-trace/
```

**Comment by t8m** [comment]
> And I can see it is set in the pkgconfig files.

That's a completely different thing.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of the 3.4 dev cycle, barring further input
